### PR TITLE
Initial support runtime dlopen in multi-threaded applications

### PR DESCRIPTION
**Pull request by sbc100**

### CODE REVIEW

File: tests/core/pthread/test_pthread_dlopen_side.c [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+// Exposing the address of `mydata` in this way forces it to
+// be present in the .data segment with it address caclulated
+// relative the `__memory_base` at which the DSO is loaded.
+int* side_data_address() {
+  return mydata;
+}
+
+// Exposing the address of `mydata` in this way forces it to
+// be present in static table region on the DSO and its
+// address will be calulated relative to the `__table_base`
```
Comment (kripken) [inline tests/core/pthread/test_pthread_dlopen_side.c:18]:
```suggestion
// address will be calculated relative to the `__table_base`
```

File: tests/core/pthread/test_pthread_dlopen_side.c [hunk 2]
```diff
@@ -0,0 +1,22 @@
+typedef int (*myfunc_type)();
+
+static int mydata[10] = { 44 };
+
+static int myfunc() {
+  return 43;
+}
+
+// Exposing the address of `mydata` in this way forces it to
+// be present in the .data segment with it address caclulated
```
Comment (kripken) [inline tests/core/pthread/test_pthread_dlopen_side.c:10]:
```suggestion
// be present in the .data segment with its address calculated
```

File: src/library_dylink.js [hunk 3]
```diff
@@ -429,17 +433,43 @@ var LibraryDylink = {
...
+      // threads already).
+      var needsAllocation = !handle || !{{{ makeGetValue('handle', C_STRUCTS.dso.mem_allocated, 'i8') }}};
+      if (needsAllocation) {
+        // alignments are powers of 2
+        var memAlign = Math.pow(2, metadata.memoryAlign);
+        // finalize alignments and verify them
+        memAlign = Math.max(memAlign, STACK_ALIGN); // we at least need stack alignment
+        // prepare memory
+        var memoryBase = metadata.memorySize ? alignMemory(getMemory(metadata.memorySize + memAlign), memAlign) : 0; // TODO: add to cleanups
+        // TODO: use only __memory_base and __table_base, need to update asm.js backend
```
Comment (kripken) [inline src/library_dylink.js:448]: asm.js backend?

**Review by kripken — APPROVED** [review]

File: src/library_dylink.js (hunk 3 above)
Comment (sbc100) [reply src/library_dylink.js:448]: Oops, that was removed in a previous PR.

**Pull request merged**